### PR TITLE
feat(earn): add function for preparing supply transaction

### PR DESCRIPTION
**Pull request by satish-ravi**
### Description

Mostly based off of swap transaction preparation

### Test plan

Unit tests, manually preparing a tx

### Related issues

- Part of ACT-1178

### Backwards compatibility

Yes

### Network scalability

If a new NetworkId and/or Network are added in the future, the changes in this PR will:

- [x] Continue to work without code changes, OR trigger a compilation error (guaranteeing we find it when a new network is added)

### CODE REVIEW

File: src/earn/prepareTransactions.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  walletAddress: Address
+  feeCurrencies: TokenBalance[]
+  poolContractAddress: Address
+}) {
+  const baseTransactions: TransactionRequest[] = []
+
+  // amount in smallest unit
+  const amountToSupply = parseUnits(amount, token.decimals)
+
+  if (!token.address) {
```
Comment (silasbw) [inline src/earn/prepareTransactions.ts:28]: i've been nagging people to use the viem type guard in places like this:

```ts
if (!token.address || !isAddress(token.address) {
```

File: src/earn/prepareTransactions.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    // should never happen
+    throw new Error('Cannot use a token without address')
+  }
+
+  const approvedAllowanceForSpender = await publicClient[
+    networkIdToNetwork[token.networkId]
+  ].readContract({
+    address: token.address as Address,
```
Comment (silasbw) [inline src/earn/prepareTransactions.ts:36]: ...then you can drop the type assertions and maybe feel good about that.

**Review by silasbw — APPROVED** [review]

File: src/earn/prepareTransactions.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  poolContractAddress: Address
+}) {
+  const baseTransactions: TransactionRequest[] = []
+
+  // amount in smallest unit
+  const amountToSupply = parseUnits(amount, token.decimals)
+
+  if (!token.address || !isAddress(token.address)) {
+    // should never happen
+    throw new Error('Cannot use a token without address')
```
Comment (MuckT) [inline src/earn/prepareTransactions.ts:42]: Could we add the tokenId to this error to help with debugging if we run into issues?

File: src/earn/prepareTransactions.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  const approvedAllowanceForSpender = await publicClient[
+    networkIdToNetwork[token.networkId]
+  ].readContract({
+    address: token.address,
+    abi: erc20.abi,
+    functionName: 'allowance',
+    args: [walletAddress, poolContractAddress],
+  })
+
+  if (approvedAllowanceForSpender < amountToSupply) {
```
Comment (MuckT) [inline src/earn/prepareTransactions.ts:54]: Would this work if `approvedAllowanceForSpender === amountToSupply`?

File: src/earn/prepareTransactions.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  if (!response.ok) {
+    throw new Error(
+      `Failed to simulate transactions. status ${response.status}, text: ${await response.text()}`
+    )
+  }
+
+  // extract fee of the supply transaction and set gas fields
+  const { simulatedTransactions }: SimulatedTransactionResponse = await response.json()
+  const supplySimulatedTx = simulatedTransactions[simulatedTransactions.length - 1]
```
Comment (MuckT) [inline src/earn/prepareTransactions.ts:100]: nit: code golf

```suggestion
  const supplySimulatedTx = simulatedTransactions.at(-1)
```

**Review by jh2oman — APPROVED** [review]

File: src/earn/prepareTransactions.ts (hunk 4 above)
Comment (satish-ravi) [reply src/earn/prepareTransactions.ts:54]: yes

File: src/earn/prepareTransactions.ts (hunk 3 above)
Comment (satish-ravi) [reply src/earn/prepareTransactions.ts]: done

File: src/earn/prepareTransactions.ts (hunk 5 above)
Comment (satish-ravi) [reply src/earn/prepareTransactions.ts:107]: this would need a type assertion, since `at` returns <array type> | undefined, so leaving as is.

**Review by MuckT — APPROVED** [review]

**Pull request merged**